src/phoenix/trace/openai/instrumentor.py

```diff
@@ -0,0 +1,279 @@
+
+
+class OpenAIInstrumentor:
+    def __init__(self, tracer: Optional[Tracer] = None) -> None:
+        """Instruments your OpenAI client to automatically create spans for each API call.
+
+        Args:
+            tracer (Optional[Tracer], optional): A tracer to record and handle spans. If not
+            provided, the default tracer will be used.
+        """
+        self._tracer = tracer or Tracer()
+
+    def instrument(self) -> None:
+        """
+        Instruments your OpenAI client.
+        """
+        openai = import_package("openai")
+        is_instrumented = hasattr(
+            openai.api_requestor.APIRequestor.request,
+            INSTRUMENTED_ATTRIBUTE_NAME,
+        )
+        if not is_instrumented:
+            openai.api_requestor.APIRequestor.request = _wrap_openai_api_requestor(
+                openai.api_requestor.APIRequestor.request, self._tracer
+            )
+            setattr(
+                openai.api_requestor.APIRequestor.request,
+                INSTRUMENTED_ATTRIBUTE_NAME,
+                True,
+            )
+
+
+def _wrap_openai_api_requestor(
+    request_fn: Callable[..., Any], tracer: Tracer
+) -> Callable[..., Any]:
+    """Wraps the OpenAI APIRequestor.request method to create spans for each API call.
+
+    Args:
+        request_fn (Callable[..., Any]): The request method on openai.api_requestor.APIRequestor.
+        tracer (Tracer): The tracer to use to create spans.
+
+    Returns:
+        Callable[..., Any]: The wrapped request method.
+    """
+
+    def wrapped(*args: Any, **kwargs: Any) -> Any:
+        call_signature = signature(request_fn)
+        bound_arguments = call_signature.bind(*args, **kwargs)
+        parameters = bound_arguments.arguments["params"]
+        is_streaming = parameters.get("stream", False)
+        url = bound_arguments.arguments["url"]
+        current_status_code = SpanStatusCode.UNSET
+        events: List[SpanEvent] = []
+        attributes: SpanAttributes = dict()
+        if not is_streaming and _get_request_type(url) is RequestType.CHAT_COMPLETION:
+            for (
+                attribute_name,
+                get_parameter_attribute_fn,
+            ) in _PARAMETER_ATTRIBUTE_FUNCTIONS.items():
+                attributes.update(
+                    _to_span_attributes(attribute_name, get_parameter_attribute_fn(parameters))
+                )
+            outputs = None
+            try:
+                start_time = datetime.now()
+                outputs = request_fn(*args, **kwargs)
+                end_time = datetime.now()
+                current_status_code = SpanStatusCode.OK
+                return outputs
+            except Exception as error:
+                end_time = datetime.now()
+                current_status_code = SpanStatusCode.ERROR
+                events.append(
+                    SpanException(
+                        message=str(error),
+                        timestamp=start_time,
```

Comment:
@RogerHYang Should we use `end_time` here?

Reply:
yea, that makes sense

Reply:
addressed in [2839654](https://github.com/Arize-ai/phoenix/pull/1550/commits/28396547c7a36fc0a23d39860842485103868791)